components/eventing-controller/pkg/backend/jetstreamv2/jetstream_internal_unit_test.go

```diff
@@ -0,0 +1,425 @@
+package jetstreamv2
+
+import (
+	"sync"
+	"testing"
+
+	backenderrors "github.com/kyma-project/kyma/components/eventing-controller/pkg/backend/jetstreamv2/errors"
+	"github.com/stretchr/testify/assert"
+
+	kymalogger "github.com/kyma-project/kyma/common/logging/logger"
+	"github.com/kyma-project/kyma/components/eventing-controller/api/v1alpha2"
+	"github.com/kyma-project/kyma/components/eventing-controller/logger"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/backend/cleaner"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/backend/metrics"
+	"github.com/kyma-project/kyma/components/eventing-controller/pkg/backend/mocks"
+	subtesting "github.com/kyma-project/kyma/components/eventing-controller/testing/v2"
+	"github.com/nats-io/nats.go"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_SyncNATSConsumers_ForEmptyTypes(t *testing.T) {
+	// given
+	callback := func(m *nats.Msg) {}
+	subWithOneType := subtesting.NewSubscriptionWithEmptyTypes()
+
+	// when
+	js := JetStream{}
+	err := js.syncNATSConsumers(subWithOneType, callback)
+
+	// then
+	assert.NoError(t, err)
+}
+
+// Test_SyncNATSConsumers tests creation/binding of the consumer and NATS Subscriptions.
+func Test_SyncNATSConsumers(t *testing.T) {
+	testEnv := getTestEnvironmentWithMock(t)
+	callback := func(m *nats.Msg) {}
+	subWithOneType := subtesting.NewSubscriptionWithOneType()
+	_, jsSubKey, defaultConsumerInfo, _ := generateConsumerInfra(testEnv.jsBackend, subWithOneType, subWithOneType.Status.Types[0])
+	subscriberMock := &mocks.Subscriber{}
+
+	testCases := []struct {
+		name                 string
+		givenSubscription    *v1alpha2.Subscription
+		givenSubscriptionMap map[SubscriptionSubjectIdentifier]Subscriber
+		givenMocks           func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier)
+		givenAssertions      func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier)
+		wantErr              error
+	}{
+		{
+			name:              "Adding a consumer/subscribing should be successful",
+			givenSubscription: subWithOneType,
+			givenMocks: func(sub *v1alpha2.Subscription, jsBackend *JetStream, jsCtx *mocks.JetStreamContext, jsSubKey SubscriptionSubjectIdentifier) {
+				for _, eventType := range sub.Status.Types {
+					_, jsSubKey, newConsumer, natsSub := generateConsumerInfra(jsBackend, sub, eventType)
+					subscribeArgs := getSubOptsAsVariadicSlice(jsBackend, sub, eventType)
+
+					jsCtx.On("ConsumerInfo", jsBackend.Config.JSStreamName, jsSubKey.ConsumerName()).Return(nil, nats.ErrConsumerNotFound)
+					jsCtx.On("AddConsumer", jsBackend.Config.JSStreamName, mock.AnythingOfType("*nats.ConsumerConfig")).Return(newConsumer, nil)
+					jsCtx.On("Subscribe", subscribeArgs...).Return(natsSub, nil)
```

Comment:
For every On() call you can just use `Test_SyncNATSConsumers` once in the table test body.

Reply:
Maybe we can get completely rid of `givenAssertions` then 🤔

Reply:
That's a very-very good point !